Feature/interpolated columns

halfnelson opened the pull request:
When trying to find an original source location for a column that doesn't line up with a segment's generated column, source tracing was returning. This PR changes the behaviour so that the lookup will use the offset between the searched for location and the segment start and apply that to the original source column.

I hit this problem when combining a sourcemap with a sourcemap that had simple line passthrough 
```js
[
[[0, 0, 0, 0]],
[[0, 0, 1, 0]]
]
```
with a more complicated transform. The behaviour implemented should give the same results as    
```js
new SourceMapConsumer(map2).originalPositionFor(new SourceMapConsumer(map1).originalPositionFor(generated))
```

jridgewell commented on .gitattributes:
```diff
@@ -0,0 +1 @@
+/test/samples/** text eol=lf
```
Thanks!

jridgewell commented on src/source-map-tree.ts:
```diff
@@ -115,22 +115,35 @@ export default class SourceMapTree {
+    // calculate the new column based on the difference between the column we
+    // were searching for, and the start of the segment we found
+    const offset = column - segment[0];
+    const originalCol = segment[3] + offset;
```
I think this requires a check that the new column is still part of this segment. Eg, if we're offseting 10 chars but the segment itself only reaches 4 chars.

halfnelson replied:
I am not sure I follow 100%. Can you think of a mapping that would cause this?, I will add a test and see

halfnelson replied:
I think I worked out what you mean. It was my misunderstanding.
I have update the code to behave like mozilla's source-map where a non exact match will return the closest lowest segments source location

jridgewell commented on src/source-map-tree.ts:
```diff
@@ -115,22 +115,32 @@ export default class SourceMapTree {
-      segment[2],
-      segment[3],
+      originalLine,
+      originalCol,
```
Nit: Can we undo these changes?

jridgewell commented on src/source-map-tree.ts:
```diff
@@ -115,22 +115,32 @@ export default class SourceMapTree {
+    if (index < 0) {
+      index = ~index - 1;
+      if (index < 0) return null; // we come before any mapped segment
+    }
```
Let's check this case first:

```suggestion
      if (index === -1) return null; // we come before any mapped segment

    // If we can't find an segment that lines up to this column, we use the
    // closest segment before it.
    if (index < 0) {
      index = ~index - 1;
    }
```

jridgewell commented on src/source-map-tree.ts:
```diff
@@ -115,22 +115,32 @@ export default class SourceMapTree {
+    }
+
     const segment = segments[index];
+    if (!segment) return null;
```
I don't believe this case is possible.

If the segments has length `n`, then `binarySearch` could return negative index down to `~n`. If we `~` that, we get the `n` back. Then your `- 1` will put us to `n - 1`. So, the last segment.

Or, if we found a match, then we'd definitely know there's a segment there.

jridgewell commented on test/samples/transpile-concat-hires/test.ts:
```diff
@@ -30,7 +30,7 @@ describe('transpile then concatenate', () => {
 
     const consumer = new SourceMapConsumer((remapped as unknown) as RawSourceMap);
-
+    // middle o in foo in bundle.js
```
I believe this is actually the `f`. Columns are 1 based, and ``` `  return 'foo';`.slice(11 - 1) ``` gives us `foo';`

To verify, we should add test cases for column `10` and `12`.

halfnelson replied:
https://github.com/mozilla/source-map/#sourcemapconsumerprototypeoriginalpositionforgeneratedposition says

 > The column number in the original source, or null if this information is not available. The column number is 0-based.

jridgewell commented on test/samples/transpile-concat-hires/test.ts:
```diff
@@ -41,6 +41,7 @@ describe('transpile then concatenate', () => {
       source: 'main.mjs',
     });
 
+    // the a in bar in bundle.js
```
Ditto for the rest of these.

jridgewell commented on test/samples/transpile-minify/test.ts:
```diff
@@ -31,7 +31,7 @@ describe('transpile then minify', () => {
     const consumer = new SourceMapConsumer((remapped as unknown) as RawSourceMap);
     const alert = consumer.originalPositionFor({
-      column: 61,
+      column: 47,
```
Ooo, I don't know how this was so wrong. But I think this should be `48`, due to the 1 base.

jridgewell commented on test/unit/source-map-tree.ts:
```diff
@@ -180,12 +185,35 @@ describe('SourceMapTree', () => {
     });
 
+    test('traces all generated cols on a line back to their source when source had characters removed', () => {
+      let expectedCols = [0, 0, 0, 0, 0, 6, 6, 6, 6];
+      let expectedLine = 4;
+      for (var genCol = 0; genCol < expectedCols.length; genCol++) {
```
```suggestion
      const expectedCols = [0, 0, 0, 0, 0, 6, 6, 6, 6];
      const expectedLine = 4;
      for (let genCol = 0; genCol < expectedCols.length; genCol++) {
```

jridgewell commented on test/unit/source-map-tree.ts:
```diff
@@ -180,12 +185,35 @@ describe('SourceMapTree', () => {
+      let expectedLine = 4;
+      for (var genCol = 0; genCol < expectedCols.length; genCol++) {
+        const trace = source.traceSegment(4, genCol, '');
+        expect(trace).toMatchObject({ line: expectedLine, column: expectedCols[genCol] });
```
Nit: We can just inline `4` for `expectedLine`.

jridgewell commented on test/unit/source-map-tree.ts:
```diff
@@ -165,9 +165,14 @@ describe('SourceMapTree', () => {
+          [0, 0, 4, 0],
+          [5, 0, 4, 6],
+        ], // line 4 is identical to line 4 of source except col 5 was removed eg 01234567890 -> 012346789
+        [[0, 0, 5, 0], [5], [6, 0, 5, 5]], // line 4 is identical to line 4 of source except a char was added at col 5 eg 01234*56789 -> 0123*456789
```
```suggestion
        [
          [0, 0, 5, 0],
          [5],
          [6, 0, 5, 5]
        ], // line 4 is identical to line 4 of source except a char was added at col 5 eg 01234*56789 -> 0123*456789
```

jridgewell commented on test/unit/source-map-tree.ts:
```diff
@@ -180,12 +185,35 @@ describe('SourceMapTree', () => {
     });
 
+    test('traces all generated cols on a line back to their source when source had characters removed', () => {
+      let expectedCols = [0, 0, 0, 0, 0, 6, 6, 6, 6];
```
Great test, now I can understand what was going wrong with the old code.

jridgewell commented on test/unit/source-map-tree.ts:
```diff
@@ -180,12 +185,35 @@ describe('SourceMapTree', () => {
+    });
+
+    test('traces all generated cols on a line back to their source when source had characters added', () => {
+      let expectedCols = [0, 0, 0, 0, 0, null, 5, 5, 5, 5, 5];
+      let expectedLine = 5;
+      for (var genCol = 0; genCol < expectedCols.length; genCol++) {
```
```suggestion
      const expectedCols = [0, 0, 0, 0, 0, null, 5, 5, 5, 5, 5];
      const expectedLine = 5;
      for (let genCol = 0; genCol < expectedCols.length; genCol++) {
```

jridgewell commented on test/unit/source-map-tree.ts:
```diff
@@ -180,12 +185,35 @@ describe('SourceMapTree', () => {
+        if (expectedCols[genCol] == null) {
+          expect(trace).toBeNull();
+        } else {
+          expect(trace).toMatchObject({ line: expectedLine, column: expectedCols[genCol] });
```
Nit: ditto inlining `expectedLine`.

halfnelson commented:
I think I have addressed all the feedback, thanks for the thoughtful review. The only changes not made are the ones that assume column is 1 based,  My reading of the documentation for source-map seems to suggest it is 0 based (while the lines are 1 based for reasons I can't currently fathom)

jridgewell commented:
Ah, you're right, it's line that's 1-based.

halfnelson commented:
Seems I had some unpushed changes, and prettier did not like the nicer formatting of that last test segment

jridgewell commented:
Everything passed? Can you open a new PR with anything you missed?
